accepted/0000-workspaces.md

```diff
@@ -0,0 +1,479 @@
+# npm workspaces
+
+## Summary
+
+Add a set of features to the npm cli that provide support to managing multiple packages from within a singular top-level, root package.
+
+## Motivation
+
+This feature has been requested by the community for a long time. The primary motivation behind this RFC is to fully realize a set of features/functionality to support managing multiple packages that may or may not be used together.
+
+The name “workspaces” is already established in the community with both Yarn and Pnpm implementing similar features under that same name so we chose to reuse it for the sake of simplicity to the larger community involved.
+
+## Detailed Explanation
+
+After sourcing feedback from the community, there are 2 major implementations/changes required in the cli in order to provide the feature set that would enable a better management of nested packages.
+- Install: In a workspaces setup users expect to be able to install all nested packages and perform the associated lifecycle scripts from the root-level, it should also be aware of sibling packages that depend on one another and link them appropriately
+- Run some npm subcommands in the context of each nested workspace
+
+## Rationale and Alternatives
+
+First and foremost there’s the alternative of leaving the problem set for userland to solve, there’s already the very popular project Lerna that provides some of these features.
+
+Also available is the alternative of supporting only the install (or bootstrap as Lerna names it) aspect of this proposal, following a less feature-rich approach but one that would still enable the basic goal of improving the user experience of managing multiple child packages but from all the feedback collected during the research phase of this RFC, this alternative is much less desirable to the community of maintainers involved.
+
+## Implementation
+
+### 1. Make npm workspace-aware
+
+We're following the lead of Yarn in supporting the `workspaces` `package.json` field which defines a list of paths, each of these paths may be a workspace itself but it also support globs.
+
+`package.json` example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
+        "packages": [
+            "packages/*"
+        ]
+    }
+}
+```
+
+`package.json` shorthand example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": [
+        "packages/*"
+    ]
+}
+```
+
+
+### 2. Installing dependencies across child packages
+
+Change `npm install` ([arborist](https://www.npmjs.com/package/@npmcli/arborist)) behavior to make it install every nested package by default once a valid configuration is defined.
+
+Arborist should also be aware of all nested workspaces and correctly link to an internal workspace should it match the required semver version of an expected dependency anywhere in the installing tree. e.g:
```

Comment:
Will Arborist also support the [`workspace:` specifier](https://pnpm.js.org/en/workspaces#workspace-ranges-workspace) supported by Yarn v2 and pnpm?

Reply:
I would also like to hear in general what the npm team thinks about these kind of additions?  This is a great way to make whole package ecosystems which have strict tool version dependence.  I would love it if we could avoid this (although yarn is already introducing these) and having a general understanding of where we all want to be in the future would be great so we can have productive conversation about these individual additions as they come up.

Reply:
> Will Arborist also support the `workspace:` specifier supported by Yarn v2 and pnpm?

no, not in this initial implementation at least - of course we need to keep sourcing feedback from the community, specially in the form of the foundation WGs (just to say that we're never shutting the door to new ideas and if the community wants any new feature outside of the scope of this RFC in the future it can always make its way into a future version)